x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/inference/adapitiveallocations/AdaptiveAllocationsScalerTests.java

```diff
@@ -0,0 +1,95 @@
+        // should scale to 2 allocations to have some spare capacity.
+        adaptiveAllocationsScaler.process(new AdaptiveAllocationsScalerService.Stats(920, 0, 0, 0.010), 10, 1);
+        assertThat(adaptiveAllocationsScaler.scale(), nullValue());
+        adaptiveAllocationsScaler.process(new AdaptiveAllocationsScalerService.Stats(950, 0, 0, 0.010), 10, 1);
+        assertThat(adaptiveAllocationsScaler.scale(), nullValue());
+        adaptiveAllocationsScaler.process(new AdaptiveAllocationsScalerService.Stats(980, 0, 0, 0.010), 10, 1);
+        assertThat(adaptiveAllocationsScaler.scale(), equalTo(2));
+        adaptiveAllocationsScaler.process(new AdaptiveAllocationsScalerService.Stats(980, 0, 0, 0.010), 10, 2);
+        assertThat(adaptiveAllocationsScaler.scale(), nullValue());
+
+        // Reducing the load to just 880 requests * 0.010 sec/request should not
+        // trigger scaling down again, to prevent oscillating.
+        adaptiveAllocationsScaler.process(new AdaptiveAllocationsScalerService.Stats(880, 0, 0, 0.010), 10, 2);
+        assertThat(adaptiveAllocationsScaler.scale(), nullValue());
+        adaptiveAllocationsScaler.process(new AdaptiveAllocationsScalerService.Stats(880, 0, 0, 0.010), 10, 2);
+        assertThat(adaptiveAllocationsScaler.scale(), nullValue());
+    }
+
+    public void testAutoscaling_respectMinMaxAllocations() {
+        AdaptiveAllocationsScaler adaptiveAllocationsScaler = new AdaptiveAllocationsScaler("test-deployment", 1);
+        adaptiveAllocationsScaler.setMinMaxNumberOfAllocations(2, 5);
+
+        // Even though there are no requests, scale to the minimum of 2 allocations.
+        adaptiveAllocationsScaler.process(new AdaptiveAllocationsScalerService.Stats(0, 0, 0, 0.010), 10, 1);
+        assertThat(adaptiveAllocationsScaler.scale(), equalTo(2));
+        adaptiveAllocationsScaler.process(new AdaptiveAllocationsScalerService.Stats(0, 0, 0, 0.010), 10, 2);
+        assertThat(adaptiveAllocationsScaler.scale(), nullValue());
+
+        // Even though there are many requests, the scale to the maximum of 5 allocations.
+        adaptiveAllocationsScaler.process(new AdaptiveAllocationsScalerService.Stats(100, 10000, 1000, 0.010), 10, 2);
+        assertThat(adaptiveAllocationsScaler.scale(), equalTo(5));
+        adaptiveAllocationsScaler.process(new AdaptiveAllocationsScalerService.Stats(500, 10000, 1000, 0.010), 10, 5);
+        assertThat(adaptiveAllocationsScaler.scale(), nullValue());
+
+        // After a while of no requests, scale to the minimum of 2 allocations.
+        adaptiveAllocationsScaler.process(new AdaptiveAllocationsScalerService.Stats(0, 0, 0, 0.010), 10, 5);
+        adaptiveAllocationsScaler.process(new AdaptiveAllocationsScalerService.Stats(0, 0, 0, 0.010), 10, 5);
+        adaptiveAllocationsScaler.process(new AdaptiveAllocationsScalerService.Stats(0, 0, 0, 0.010), 10, 5);
+        assertThat(adaptiveAllocationsScaler.scale(), equalTo(2));
+    }
```

Comment:
As before I would like to stress test the distribution assumptions for the inference duration.

Reply:
Do you have a concrete suggestion for a test?

Reply:
So I would generate inference durations from a distribution whose variance >> mean, so something like log-normal(0.1, 1.5) would work. Now you reminded me we don't detect manoeuvres for inference time I'm less worried. However, it would perhaps be worth checking there aren't pathologies.

It might also be interesting to test behaviour doing the same for request rates, i.e. generate bucket counts using the same distribution. Since the Poisson process assumption is reasonable, I'm also less concerned about this case, but again, it might be interesting to see how it does behave.